dev/bus/export/dev/serial.hpp

```diff
@@ -0,0 +1,217 @@
+uint8_t serial<PBus, buf_size>::m_tx_buffer[buf_size];
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_write_iter;
+
+template <class PBus, size_t buf_size>
+std::atomic_size_t serial<PBus, buf_size>::m_rx_read_iter;
+
+template <class PBus, size_t buf_size>
+bool serial<PBus, buf_size>::m_is_inited;
+
+template <class PBus, size_t buf_size>
+binary_semaphore serial<PBus, buf_size>::m_rx_new_xfer_allowed;
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::init()
+{
+    ecl_assert(m_is_inited == false);
+    auto result = PBus::init();
+    if (is_error(result)) {
+        return result;
+    }
+    PBus::set_handler(bus_handler);
+    result = PBus::enable_listen_mode();
+    if (is_error(result)) {
+        return result;
+    }
+    m_rx_new_xfer_allowed.signal();
+    PBus::set_rx(m_rx_buffer, buffer_size);
+    result = PBus::do_xfer();
+    if (is_ok(result)) {
+        m_is_inited = true;
+    }
+    return result;
+}
+
+template <class PBus, size_t buf_size>
+void serial<PBus, buf_size>::bus_handler(bus_channel ch,
+        bus_event type, size_t total)
+{
+    if (ch == bus_channel::rx) {
+        ecl_assert(type == bus_event::tc);
+        m_rx_write_iter++;
+        ecl_assert(m_rx_write_iter == total);
+
+        // Start new xfer if needed and allowed
+        if (m_rx_new_xfer_allowed.try_wait()) {
+            try_start_xfer();
+            // Allow user to read from rx_buffer
+            m_rx_new_xfer_allowed.signal();
+        }
+    }
+    // TODO: handle TX events
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_byte(uint8_t &byte)
+{
+    size_t sz = 1;
+    return recv_buf(&byte, sz);
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::recv_buf(uint8_t *buf, size_t &sz)
+{
+    ecl_assert(m_is_inited == true);
+    // Restrict starting new xfer from bus_handler()
+    // to prevent data loss
+    m_rx_new_xfer_allowed.wait();
+
+    // Wait until data is available on the read end
+    while ((m_rx_read_iter == m_rx_write_iter)) { }
+
+    sz = std::min(sz, m_rx_write_iter - m_rx_read_iter);
+    std::copy(m_rx_buffer + m_rx_read_iter,
+            m_rx_buffer + m_rx_read_iter + sz, buf);
+    m_rx_read_iter += sz;
+
+    // Start new xfer if required
+    auto result = try_start_xfer();
+
+    // Wrapping buffer iterator should be done after try_start_xfer()
+    if (m_rx_read_iter == buffer_size) {
+        m_rx_read_iter = 0;
+    }
+    // Allow starting new xfer from bus_handler()
+    m_rx_new_xfer_allowed.signal();
+    return result;
+}
+
+template <class PBus, size_t buf_size>
+err serial<PBus, buf_size>::try_start_xfer()
+{
+    err result = err::ok;
+    if (m_rx_write_iter == buffer_size) {
+        m_rx_write_iter = 0;
+        // Notify caller that some unread data might be lost
+        if (m_rx_read_iter < buffer_size) {
+            result = err::nobufs;
+        }
+        m_rx_read_iter = 0;
```

Comment:
Suppose that `m_rx_read_iter = buffer_size /  2` and `m_rx_write_iter == buffer_size`. If so, half of the buffer will be lost.